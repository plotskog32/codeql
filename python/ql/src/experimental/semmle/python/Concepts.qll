/**
 * Provides abstract classes representing generic concepts such as file system
 * access or system command execution, for which individual framework libraries
 * provide concrete subclasses.
 */

import python
private import experimental.dataflow.DataFlow
private import experimental.semmle.python.Frameworks

/**
 * A data-flow node that executes an operating system command,
 * for instance by spawning a new process.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `SystemCommandExecution::Range` instead.
 */
class SystemCommandExecution extends DataFlow::Node {
  SystemCommandExecution::Range self;

  SystemCommandExecution() { this = self }

  /** Gets the argument that specifies the command to be executed. */
  DataFlow::Node getCommand() { result = self.getCommand() }
}

/** Provides a class for modeling new system-command execution APIs. */
module SystemCommandExecution {
  /**
   * A data-flow node that executes an operating system command,
   * for instance by spawning a new process.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `SystemCommandExecution` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Gets the argument that specifies the command to be executed. */
    abstract DataFlow::Node getCommand();
  }
}

/**
 * A data-flow node that decodes data from a binary or textual format. This
 * is intended to include deserialization, unmarshalling, decoding, unpickling,
 * decompressing, decrypting, parsing etc.
 *
 * Doing so should normally preserve taint, but it can also be a problem
 * in itself, e.g. if it allows code execution or could result in deinal-of-service.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `Decoding::Range` instead.
 */
class Decoding extends DataFlow::Node {
  Decoding::Range self;

  Decoding() { this = self }

  /** Holds if this call is unsafe, e.g. if it may execute arbitrary code. */
  predicate unsafe() { self.unsafe() }

  /** Gets an input that is decoded by this function. */
  DataFlow::Node getAnInput() { result = self.getAnInput() }

  /** Gets the output that contains the decoded data produced by this function. */
  DataFlow::Node getOutput() { result = self.getOutput() }

  /** Gets an identifier for the format this function decodes from, such as "JSON". */
  string getFormat() { result = self.getFormat() }
}

/** Provides a class for modeling new decoding mechanisms. */
module Decoding {
  /**
   * A data-flow node that decodes data from a binary or textual format. This
   * is intended to include deserialization, unmarshalling, decoding, unpickling,
   * decompressing, decrypting, parsing etc.
   *
   * Doing so should normally preserve taint, but it can also be a problem
   * in itself, e.g. if it allows code execution or could result in deinal-of-service.
   *
   * Extend this class to model new APIs. If you want to refine existing API models,
   * extend `Decoding` instead.
   */
  abstract class Range extends DataFlow::Node {
    /** Holds if this call is unsafe, e.g. if it may execute arbitrary code. */
    abstract predicate unsafe();

    /** Gets an input that is decoded by this function. */
    abstract DataFlow::Node getAnInput();

    /** Gets the output that contains the decoded data produced by this function. */
    abstract DataFlow::Node getOutput();

    /** Gets an identifier for the format this function decodes from, such as "JSON". */
    abstract string getFormat();
  }
}
