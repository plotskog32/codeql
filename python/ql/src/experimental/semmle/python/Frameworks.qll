/**
 * Helper file that imports all framework modeling.
 */

private import experimental.semmle.python.frameworks.Flask
private import experimental.semmle.python.frameworks.Django
private import experimental.semmle.python.frameworks.Marshal
private import experimental.semmle.python.frameworks.Pickle
private import experimental.semmle.python.frameworks.Stdlib
private import experimental.semmle.python.frameworks.Yaml
