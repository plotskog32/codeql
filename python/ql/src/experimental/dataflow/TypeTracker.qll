/** Step Summaries and Type Tracking */

import python
import internal.DataFlowPublic
import internal.DataFlowPrivate

/** Any string that may appear as the name of an attribute or access path. */
class AttributeName extends string {
  AttributeName() { this = any(Attribute a).getName() }
}

/** Either an attribute name, or the empty string (representing no attribute). */
class OptionalAttributeName extends string {
  OptionalAttributeName() { this instanceof AttributeName or this = "" }
}

/**
 * A description of a step on an inter-procedural data flow path.
 */
private newtype TStepSummary =
  LevelStep() or
  CallStep() or
  ReturnStep() or
  StoreStep(AttributeName attr) or
  LoadStep(AttributeName attr) or
  CopyStep(AttributeName attr)

/**
 * INTERNAL: Use `TypeTracker` or `TypeBackTracker` instead.
 *
 * A description of a step on an inter-procedural data flow path.
 */
class StepSummary extends TStepSummary {
  /** Gets a textual representation of this step summary. */
  string toString() {
    this instanceof LevelStep and result = "level"
    or
    this instanceof CallStep and result = "call"
    or
    this instanceof ReturnStep and result = "return"
    or
    exists(string prop | this = StoreStep(prop) | result = "store " + prop)
    or
    exists(string prop | this = LoadStep(prop) | result = "load " + prop)
  }
}

module StepSummary {
  cached
  predicate step(Node pred, Node succ, StepSummary summary) {
    exists(Node mid | EssaFlow::essaFlowStep*(pred, mid) and smallstep(mid, succ, summary))
  }

  predicate smallstep(Node pred, Node succ, StepSummary summary) {
    EssaFlow::essaFlowStep(pred, succ) and
    summary = LevelStep()
    or
    callStep(pred, succ) and summary = CallStep()
    or
    returnStep(pred, succ) and
    summary = ReturnStep()
    or
    exists(string attr |
      basicStoreStep(pred, succ, attr) and
      summary = StoreStep(attr)
      or
      basicLoadStep(pred, succ, attr) and summary = LoadStep(attr)
    )
  }
}

/** Holds if `pred` steps to `succ` by being passed as a parameter in a call. */
predicate callStep(ArgumentNode pred, ParameterNode succ) {
  // TODO: Support special methods?
  exists(DataFlowCall call, int i |
    pred.argumentOf(call, i) and succ.isParameterOf(call.getCallable(), i)
  )
}

/** Holds if `pred` steps to `succ` by being returned from a call. */
predicate returnStep(ReturnNode pred, Node succ) {
  exists(DataFlowCall call |
    pred.getEnclosingCallable() = call.getCallable() and succ.asCfgNode() = call
  )
}

/**
 * Holds if `pred` is being written to the `attr` attribute of the object in `succ`.
 *
 * Note that the choice of `succ` does not have to make sense "chronologically".
 * All we care about is whether the `attr` attribute of `succ` can have a specific type,
 * and the assumption is that if a specific type appears here, then any access of that
 * particular attribute can yield something of that particular type.
 *
 * Thus, in an example such as
 *
 * ```python
 * def foo(y):
 *    x = Foo()
 *    bar(x)
 *    x.attr = y
 *    baz(x)
 *
 * def bar(x):
 *    z = x.attr
 * ```
 * for the attribute write `x.attr = y`, we will have `attr` being the literal string `"attr"`,
 * `pred` will be `y`, and `succ` will be the object `Foo()` created on the first line of the
 * function. This means we will track the fact that `x.attr` can have the type of `y` into the
 * assignment to `z` inside `bar`, even though this attribute write happens _after_ `bar` is called.
 */
predicate basicStoreStep(Node pred, Node succ, string attr) {
  exists(AttributeAssignment a, Node var |
    a.getName() = attr and
    EssaFlow::essaFlowStep*(succ, var) and
    var.asVar() = a.getInput() and
    pred.asCfgNode() = a.getValue()
  )
}

/**
 * Holds if `succ` is the result of accessing the `attr` attribute of `pred`.
 */
predicate basicLoadStep(Node pred, Node succ, string attr) {
  exists(AttrNode s | succ.asCfgNode() = s and s.getObject(attr) = pred.asCfgNode())
}

/**
 * A utility class that is equivalent to `boolean` but does not require type joining.
 */
private class Boolean extends boolean {
  Boolean() { this = true or this = false }
}

private newtype TTypeTracker = MkTypeTracker(Boolean hasCall, OptionalAttributeName prop)

/**
 * Summary of the steps needed to track a value to a given dataflow node.
 *
 * This can be used to track objects that implement a certain API in order to
 * recognize calls to that API. Note that type-tracking does not by itself provide a
 * source/sink relation, that is, it may determine that a node has a given type,
 * but it won't determine where that type came from.
 *
 * It is recommended that all uses of this type are written in the following form,
 * for tracking some type `myType`:
 * ```
 * Node myType(DataFlow::TypeTracker t) {
 *   t.start() and
 *   result = < source of myType >
 *   or
 *   exists (TypeTracker t2 |
 *     result = myType(t2).track(t2, t)
 *   )
 * }
 *
 * DataFlow::SourceNode myType() { result = myType(DataFlow::TypeTracker::end()) }
 * ```
 *
 * Instead of `result = myType(t2).track(t2, t)`, you can also use the equivalent
 * `t = t2.step(myType(t2), result)`. If you additionally want to track individual
 * intra-procedural steps, use `t = t2.smallstep(myCallback(t2), result)`.
 */
class TypeTracker extends TTypeTracker {
  Boolean hasCall;
  OptionalAttributeName prop;

  TypeTracker() { this = MkTypeTracker(hasCall, prop) }

  /** Gets the summary resulting from appending `step` to this type-tracking summary. */
  cached
  TypeTracker append(StepSummary step) {
    step = LevelStep() and result = this
    or
    step = CallStep() and result = MkTypeTracker(true, prop)
    or
    step = ReturnStep() and hasCall = false and result = this
    or
    step = LoadStep(prop) and result = MkTypeTracker(hasCall, "")
    or
    exists(string p | step = StoreStep(p) and prop = "" and result = MkTypeTracker(hasCall, p))
  }

  /** Gets a textual representation of this summary. */
  string toString() {
    exists(string withCall, string withProp |
      (if hasCall = true then withCall = "with" else withCall = "without") and
      (if prop != "" then withProp = " with property " + prop else withProp = "") and
      result = "type tracker " + withCall + " call steps" + withProp
    )
  }

  /**
   * Holds if this is the starting point of type tracking.
   */
  predicate start() { hasCall = false and prop = "" }

  /**
   * Holds if this is the starting point of type tracking, and the value starts in the property named `propName`.
   * The type tracking only ends after the property has been loaded.
   */
  predicate startInProp(AttributeName propName) { hasCall = false and prop = propName }

  /**
   * Holds if this is the starting point of type tracking
   * when tracking a parameter into a call, but not out of it.
   */
  predicate call() { hasCall = true and prop = "" }

  /**
   * Holds if this is the end point of type tracking.
   */
  predicate end() { prop = "" }

  /**
   * INTERNAL. DO NOT USE.
   *
   * Holds if this type has been tracked into a call.
   */
  boolean hasCall() { result = hasCall }

  /**
   * INTERNAL. DO NOT USE.
   *
   * Gets the property associated with this type tracker.
   */
  string getProp() { result = prop }

  /**
   * Gets a type tracker that starts where this one has left off to allow continued
   * tracking.
   *
   * This predicate is only defined if the type has not been tracked into a property.
   */
  TypeTracker continue() { prop = "" and result = this }

  /**
   * Gets the summary that corresponds to having taken a forwards
   * heap and/or inter-procedural step from `pred` to `succ`.
   */
  pragma[inline]
  TypeTracker step(Node pred, Node succ) {
    exists(StepSummary summary |
      StepSummary::step(pred, succ, summary) and
      result = this.append(summary)
    )
  }

  /**
   * Gets the summary that corresponds to having taken a forwards
   * local, heap and/or inter-procedural step from `pred` to `succ`.
   *
   * Unlike `TypeTracker::step`, this predicate exposes all edges
   * in the flow graph, and not just the edges between `SourceNode`s.
   * It may therefore be less performant.
   *
   * Type tracking predicates using small steps typically take the following form:
   * ```ql
   * DataFlow::Node myType(DataFlow::TypeTracker t) {
   *   t.start() and
   *   result = < source of myType >
   *   or
   *   exists (DataFlow::TypeTracker t2 |
   *     t = t2.smallstep(myType(t2), result)
   *   )
   * }
   *
   * DataFlow::Node myType() {
   *   result = myType(DataFlow::TypeTracker::end())
   * }
   * ```
   */
  pragma[inline]
  TypeTracker smallstep(Node pred, Node succ) {
    exists(StepSummary summary |
      StepSummary::smallstep(pred, succ, summary) and
      result = this.append(summary)
    )
    or
    EssaFlow::essaFlowStep(pred, succ) and
    result = this
  }
}
