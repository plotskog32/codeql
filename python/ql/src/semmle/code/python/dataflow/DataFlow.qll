/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) data flow analyses.
 */

import python

module DataFlow {
  import semmle.code.python.dataflow.internal.DataFlowImpl
}
