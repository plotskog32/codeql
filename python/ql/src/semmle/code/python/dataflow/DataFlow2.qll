/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) data flow analyses.
 */

import python

module DataFlow2 {
  import semmle.code.python.dataflow.internal.DataFlowImpl2
}
