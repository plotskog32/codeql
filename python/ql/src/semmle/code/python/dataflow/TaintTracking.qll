/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) taint-tracking analyses.
 */

import python

module TaintTracking {
  import semmle.code.python.dataflow.internal.tainttracking1.TaintTrackingImpl
}