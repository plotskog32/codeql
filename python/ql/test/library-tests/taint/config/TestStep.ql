import python
import semmle.python.security.TaintTracking
import TaintLib
import semmle.python.dataflow.Implementation


from TaintTrackingNode n, TaintTrackingNode s, TaintTracking::Configuration config
where s = n.getASuccessor() and config = n.getConfiguration()
select 
    config + ":", 
    n.getTaintKind(), n.getLocation().toString(), n.getNode().toString(), n.getContext(),
    " --> ",
    s.getTaintKind(), s.getLocation().toString(), s.getNode().toString(), s.getContext()
