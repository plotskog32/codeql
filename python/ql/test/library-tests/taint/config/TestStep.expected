| Basic custom config: | simple.test | test.py:120 | CUSTOM_SOURCE |  |  -->  | simple.test | test.py:121 | t |  |
| Basic custom config: | simple.test | test.py:126 | CUSTOM_SOURCE |  |  -->  | simple.test | test.py:130 | t |  |
| Basic custom config: | simple.test | test.py:136 | CUSTOM_SOURCE |  |  -->  | simple.test | test.py:142 | t |  |
| Basic custom config: | simple.test | test.py:146 | CUSTOM_SOURCE |  |  -->  | simple.test | test.py:149 | t |  |
| Rock-paper-scissors config: | paper | rockpaperscissors.py:25 | Attribute() |  |  -->  | paper | rockpaperscissors.py:26 | y |  |
| Rock-paper-scissors config: | paper | rockpaperscissors.py:26 | y |  |  -->  | paper | rockpaperscissors.py:9 | arg | Parameter 0(no attribute) is paper |
| Rock-paper-scissors config: | paper | rockpaperscissors.py:30 | Attribute() |  |  -->  | paper | rockpaperscissors.py:32 | y |  |
| Rock-paper-scissors config: | paper | rockpaperscissors.py:32 | y |  |  -->  | paper | rockpaperscissors.py:6 | arg | Parameter 0(no attribute) is paper |
| Rock-paper-scissors config: | rock | rockpaperscissors.py:16 | ROCK |  |  -->  | rock | rockpaperscissors.py:6 | arg | Parameter 0(no attribute) is rock |
| Rock-paper-scissors config: | rock | rockpaperscissors.py:19 | ROCK |  |  -->  | rock | rockpaperscissors.py:20 | x |  |
| Rock-paper-scissors config: | rock | rockpaperscissors.py:20 | x |  |  -->  | scissors | rockpaperscissors.py:20 | Attribute() |  |
| Rock-paper-scissors config: | rock | rockpaperscissors.py:24 | ROCK |  |  -->  | rock | rockpaperscissors.py:25 | x |  |
| Rock-paper-scissors config: | rock | rockpaperscissors.py:25 | x |  |  -->  | scissors | rockpaperscissors.py:25 | Attribute() |  |
| Rock-paper-scissors config: | scissors | rockpaperscissors.py:13 | SCISSORS |  |  -->  | scissors | rockpaperscissors.py:3 | arg | Parameter 0(no attribute) is scissors |
| Rock-paper-scissors config: | scissors | rockpaperscissors.py:20 | Attribute() |  |  -->  | scissors | rockpaperscissors.py:21 | y |  |
| Rock-paper-scissors config: | scissors | rockpaperscissors.py:21 | y |  |  -->  | scissors | rockpaperscissors.py:9 | arg | Parameter 0(no attribute) is scissors |
| Rock-paper-scissors config: | scissors | rockpaperscissors.py:25 | Attribute() |  |  -->  | paper | rockpaperscissors.py:25 | Attribute() |  |
| Rock-paper-scissors config: | scissors | rockpaperscissors.py:29 | SCISSORS |  |  -->  | scissors | rockpaperscissors.py:30 | x |  |
| Rock-paper-scissors config: | scissors | rockpaperscissors.py:29 | SCISSORS |  |  -->  | scissors | rockpaperscissors.py:31 | x |  |
| Rock-paper-scissors config: | scissors | rockpaperscissors.py:30 | x |  |  -->  | paper | rockpaperscissors.py:30 | Attribute() |  |
| Rock-paper-scissors config: | scissors | rockpaperscissors.py:31 | x |  |  -->  | scissors | rockpaperscissors.py:6 | arg | Parameter 0(no attribute) is scissors |
| Simple config: | [simple.test] | test.py:168 | List |  |  -->  | [simple.test] | test.py:170 | l |  |
| Simple config: | [simple.test] | test.py:168 | List |  |  -->  | [simple.test] | test.py:174 | l |  |
| Simple config: | [simple.test] | test.py:174 | l |  |  -->  | [simple.test] | test.py:174 | list() |  |
| Simple config: | [simple.test] | test.py:208 | List |  |  -->  | [simple.test] | test.py:209 | seq |  |
| Simple config: | [simple.test] | test.py:209 | seq |  |  -->  | simple.test | test.py:209 | For |  |
| Simple config: | [simple.test] | test.py:213 | flow_in_generator() |  |  -->  | simple.test | test.py:213 | For |  |
| Simple config: | iterable.simple | test.py:213 | flow_in_generator() |  |  -->  | simple.test | test.py:213 | For |  |
| Simple config: | simple.test | carrier.py:4 | arg | Parameter 1(no attribute) is simple.test |  -->  | simple.test | carrier.py:5 | arg | Parameter 1(no attribute) is simple.test |
| Simple config: | simple.test | carrier.py:17 | SOURCE |  |  -->  | simple.test | carrier.py:4 | arg | Parameter 1(no attribute) is simple.test |
| Simple config: | simple.test | carrier.py:25 | SOURCE |  |  -->  | simple.test | carrier.py:4 | arg | Parameter 1(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:2 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:3 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:3 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:6 | f1() | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:5 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:6 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:6 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:2 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:6 | f1() | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:9 | f2() | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:8 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:9 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:9 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:5 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:9 | f2() | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:12 | f3() | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:11 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:12 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:12 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:8 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:12 | f3() | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:15 | f4() | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:14 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:15 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:15 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:11 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:15 | f4() | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:18 | f5() | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:17 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:18 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:18 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:14 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:18 | f5() | Parameter 0(no attribute) is simple.test |  -->  | simple.test | deep.py:20 | f6() |  |
| Simple config: | simple.test | deep.py:20 | SOURCE |  |  -->  | simple.test | deep.py:17 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | deep.py:20 | f6() |  |  -->  | simple.test | deep.py:22 | x |  |
| Simple config: | simple.test | module.py:3 | SOURCE |  |  -->  | simple.test | test.py:155 | ImportMember |  |
| Simple config: | simple.test | test.py:6 | SOURCE |  |  -->  | simple.test | test.py:7 | s |  |
| Simple config: | simple.test | test.py:12 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | test.py:13 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | test.py:20 | SOURCE |  |  -->  | simple.test | test.py:21 | t |  |
| Simple config: | simple.test | test.py:21 | t |  |  -->  | simple.test | test.py:12 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | test.py:37 | SOURCE |  |  -->  | simple.test | test.py:41 | t |  |
| Simple config: | simple.test | test.py:49 | arg | Parameter 1(no attribute) is simple.test |  -->  | simple.test | test.py:51 | arg | Parameter 1(no attribute) is simple.test |
| Simple config: | simple.test | test.py:51 | arg | Parameter 1(no attribute) is simple.test |  -->  | simple.test | test.py:12 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | test.py:62 | SOURCE |  |  -->  | simple.test | test.py:63 | t |  |
| Simple config: | simple.test | test.py:63 | t |  |  -->  | simple.test | test.py:49 | arg | Parameter 1(no attribute) is simple.test |
| Simple config: | simple.test | test.py:67 | SOURCE |  |  -->  | simple.test | test.py:70 | t |  |
| Simple config: | simple.test | test.py:70 | t |  |  -->  | simple.test | test.py:49 | arg | Parameter 1(no attribute) is simple.test |
| Simple config: | simple.test | test.py:72 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | test.py:73 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | test.py:73 | arg | Parameter 0(no attribute) is simple.test |  -->  | simple.test | test.py:77 | hub() |  |
| Simple config: | simple.test | test.py:76 | SOURCE |  |  -->  | simple.test | test.py:77 | t |  |
| Simple config: | simple.test | test.py:77 | hub() |  |  -->  | simple.test | test.py:78 | t |  |
| Simple config: | simple.test | test.py:77 | t |  |  -->  | simple.test | test.py:72 | arg | Parameter 0(no attribute) is simple.test |
| Simple config: | simple.test | test.py:128 | SOURCE |  |  -->  | simple.test | test.py:132 | t |  |
| Simple config: | simple.test | test.py:138 | SOURCE |  |  -->  | simple.test | test.py:140 | t |  |
| Simple config: | simple.test | test.py:148 | SOURCE |  |  -->  | simple.test | test.py:149 | t |  |
| Simple config: | simple.test | test.py:155 | ImportMember |  |  -->  | simple.test | test.py:156 | unsafe |  |
| Simple config: | simple.test | test.py:163 | SOURCE |  |  -->  | simple.test | test.py:164 | s |  |
| Simple config: | simple.test | test.py:163 | SOURCE |  |  -->  | simple.test | test.py:165 | s |  |
| Simple config: | simple.test | test.py:168 | SOURCE |  |  -->  | [simple.test] | test.py:168 | List |  |
| Simple config: | simple.test | test.py:169 | SOURCE |  |  -->  | {simple.test} | test.py:169 | Dict |  |
| Simple config: | simple.test | test.py:178 | SOURCE |  |  -->  | simple.test | test.py:179 | t |  |
| Simple config: | simple.test | test.py:178 | SOURCE |  |  -->  | simple.test | test.py:180 | t |  |
| Simple config: | simple.test | test.py:178 | SOURCE |  |  -->  | simple.test | test.py:183 | t |  |
| Simple config: | simple.test | test.py:178 | SOURCE |  |  -->  | simple.test | test.py:186 | t |  |
| Simple config: | simple.test | test.py:195 | SOURCE |  |  -->  | simple.test | test.py:196 | t |  |
| Simple config: | simple.test | test.py:195 | SOURCE |  |  -->  | simple.test | test.py:199 | t |  |
| Simple config: | simple.test | test.py:208 | SOURCE |  |  -->  | [simple.test] | test.py:208 | List |  |
| Simple config: | simple.test | test.py:209 | For |  |  -->  | simple.test | test.py:210 | i |  |
| Simple config: | simple.test | test.py:210 | i |  |  -->  | [simple.test] | test.py:213 | flow_in_generator() |  |
| Simple config: | simple.test | test.py:210 | i |  |  -->  | iterable.simple | test.py:213 | flow_in_generator() |  |
| Simple config: | simple.test | test.py:213 | For |  |  -->  | simple.test | test.py:214 | x |  |
| Simple config: | {simple.test} | test.py:169 | Dict |  |  -->  | {simple.test} | test.py:171 | d |  |
| Simple config: | {simple.test} | test.py:169 | Dict |  |  -->  | {simple.test} | test.py:175 | d |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {Command injection} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {SQL injection} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[Command injection]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[SQL injection]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[basic.custom]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[explicit.carrier]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[falsey]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[iterable.simple]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[paper]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[rock]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[scissors]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {[simple.test]} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {basic.custom} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {explicit.carrier} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {falsey} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {iterable.simple} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {paper} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {rock} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {scissors} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {simple.test} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{Command injection}} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{SQL injection}} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{basic.custom}} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{explicit.carrier}} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{falsey}} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{iterable.simple}} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{paper}} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{rock}} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{scissors}} | test.py:175 | dict() |  |
| Simple config: | {simple.test} | test.py:175 | d |  |  -->  | {{simple.test}} | test.py:175 | dict() |  |
| Taint carrier config: | explicit.carrier | carrier.py:4 | arg | Parameter 1(no attribute) is explicit.carrier |  -->  | explicit.carrier | carrier.py:5 | arg | Parameter 1(no attribute) is explicit.carrier |
| Taint carrier config: | explicit.carrier | carrier.py:13 | arg | Parameter 0(no attribute) is explicit.carrier |  -->  | explicit.carrier | carrier.py:14 | arg | Parameter 0(no attribute) is explicit.carrier |
| Taint carrier config: | explicit.carrier | carrier.py:14 | arg | Parameter 0(no attribute) is explicit.carrier |  -->  | explicit.carrier | carrier.py:29 | hub() |  |
| Taint carrier config: | explicit.carrier | carrier.py:21 | TAINT_CARRIER_SOURCE |  |  -->  | explicit.carrier | carrier.py:22 | c |  |
| Taint carrier config: | explicit.carrier | carrier.py:22 | c |  |  -->  | simple.test | carrier.py:22 | Attribute() |  |
| Taint carrier config: | explicit.carrier | carrier.py:29 | TAINT_CARRIER_SOURCE |  |  -->  | explicit.carrier | carrier.py:13 | arg | Parameter 0(no attribute) is explicit.carrier |
| Taint carrier config: | explicit.carrier | carrier.py:29 | hub() |  |  -->  | explicit.carrier | carrier.py:30 | c |  |
| Taint carrier config: | explicit.carrier | carrier.py:30 | c |  |  -->  | simple.test | carrier.py:30 | Attribute() |  |
| Taint carrier config: | explicit.carrier | carrier.py:33 | TAINT_CARRIER_SOURCE |  |  -->  | explicit.carrier | carrier.py:4 | arg | Parameter 1(no attribute) is explicit.carrier |
